File: packages/core/src/mulliken.ts
import type { PitchClass } from './types';
import { transpose, areEqual } from './pcset';
import { inversionalAxes } from './symmetry';
import { isRetrogradePalindrome } from './modes';

export function mullikenLabel(pcs: PitchClass[]): string {
  const t6Symmetric = areEqual(transpose(pcs, 6), pcs);
  const hasInversion = inversionalAxes(pcs).length > 0;
  const palindrome = isRetrogradePalindrome(pcs);

  const primary = t6Symmetric ? 'A' : 'B';
  const subscript = hasInversion ? '1' : '2';
  const parity = palindrome ? 'g' : 'u';

  return `${primary}${subscript}${parity}`;
}

File: packages/core-swift/Sources/ChrometriaCore/Mulliken.swift
import Foundation

/// Mulliken symmetry label (e.g. `"B2u"`), mirroring
/// `packages/core/src/mulliken.ts` exactly.
///
/// - Primary (`A`/`B`): whether the set is invariant under T6.
/// - Subscript (`1`/`2`): whether the set has any inversional symmetry.
/// - Parity (`g`/`u`): whether the set's adjacent-interval pattern is a
///   retrograde palindrome.
public func mullikenLabel(_ pcs: [PitchClass]) -> String {
    let t6Symmetric = areEqual(transpose(pcs, 6), pcs)
    let hasInversion = !inversionalAxes(pcs).isEmpty
    let palindrome = isRetrogradePalindrome(pcs)

    let primary = t6Symmetric ? "A" : "B"
    let subscriptChar = hasInversion ? "1" : "2"
    let parity = palindrome ? "g" : "u"

    return "\(primary)\(subscriptChar)\(parity)"
}
